docker-compose initial setup

agowa opened the pull request:
As discussed in https://github.com/pretix/pretix/pull/1068
For PostgreSQL, don't merge now, yet.

I'll add it later today, but have to test the setup first before I push it.

agowa committed:
Initial commit

agowa committed:
Add pretix as submodule

agowa committed:
Add postgresql, settings file and update README.md
Fixes #1

agowa commented:
Ok, now with postgresql, documentation and centralized configuration file.

oocf commented on Dockerfile:
```diff
@@ -0,0 +1,45 @@
+	&& /usr/sbin/update-locale LANG=C.UTF-8
+RUN mkdir /etc/pretix /data /static
+RUN pip3 install -U pip wheel setuptools uwsgi
+
```
I think you could use the same RUN command for lines 16, 23, 26 and 27
and you should reduce the docker container size

agowa replied:
That will make the build cache unusable.
And this would also not reduce the size, as nothing gets deleted, but added here.

oocf commented on postfix/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+    && rm -rf /var/lib/apt/lists/*
+RUN sed -i 's_^mynetworks = .*$_mynetworks = 0.0.0.0/0 [::]/0_g' /etc/postfix/main.cf
+RUN mkdir -p /var/spool/postfix/etc/ssl && chown -R postfix /var/spool/postfix
+
```
Same here you could use the same RUN command in the same line for lines 5, 11 and 12

agowa replied:
Have your read this guide? https://docs.docker.com/develop/develop-images/dockerfile_best-practices
Having multiple layers is not bad. It is only bad, if you try to delete things in the new layer, as that does not shrink the size.
If you don't believe me, you could use [dive](https://github.com/wagoodman/dive) or [docker-explorer](https://github.com/google/docker-explorer) instead.

oocf commented on nginx/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+    addgroup -S www-data && \
+    adduser -H -h /static -S -G www-data -s /bin/false www-data
+RUN rm -rf /etc/nginx/conf.d/*
+ADD nginx.conf /etc/nginx/nginx.conf
```
Same here add a single line for the RUN in line 3 and 7

agowa replied:
Normally the "rm -rf" should be part of the same layer that adds those files. But this is a special case, as these files are already within the base image "nginx:alpine", so we cannot delete them from the image at all, but only "mark them as deleted" in a new layer.
combining it with the former layer does therefore not make sense. One could argue, that it should be in front of the software update/installation, because it removes things from the base image and the software update step is expected to change more frequently than the rm or the base image. Therefore the update should be further down, to better utilize layer caching for building images. But it does not really matter to the user that much...

oocf reviewed (changes requested):
Use the RUN command for all the different commands that RUN one after another so you have less layer and your docker container should be smaller.

Commit:
Add healthcheck to nginx

agowa committed:
Update to latest nginx container (nginx:alpine now includes the www-data group by default)

agowa committed:
Update pretix project reference to use https url and reference master

Commit:
ngnix.restart_policy.condition any => on-failure

services.ngnix.restart_policy.condition any => on-failure

snyk-bot committed:
fix: Dockerfile to reduce vulnerabilities

The following vulnerabilities are fixed with an upgrade:
- https://snyk.io/vuln/SNYK-DEBIAN9-GLIBC-356851
- https://snyk.io/vuln/SNYK-DEBIAN9-GLIBC-356851
- https://snyk.io/vuln/SNYK-DEBIAN9-IMAGEMAGICK-1244103
- https://snyk.io/vuln/SNYK-DEBIAN9-IMAGEMAGICK-1246265
- https://snyk.io/vuln/SNYK-DEBIAN9-IMAGEMAGICK-1246515

snyk-bot committed:
fix: Dockerfile to reduce vulnerabilities

The following vulnerabilities are fixed with an upgrade:
- https://snyk.io/vuln/SNYK-DEBIAN10-LIBGCRYPT20-1297893
- https://snyk.io/vuln/SNYK-DEBIAN10-SYSTEMD-345386
- https://snyk.io/vuln/SNYK-DEBIAN10-SYSTEMD-345386
- https://snyk.io/vuln/SNYK-DEBIAN10-SYSTEMD-345391
- https://snyk.io/vuln/SNYK-DEBIAN10-SYSTEMD-345391
